src/base/memory_monitor_sr.cpp

```diff
@@ -0,0 +1,235 @@
+{
+  memory_monitor *mmon_Gl = nullptr;
+
+  const char *mmon_subcomponent::get_name ()
+  {
+    return m_subcomp_name.c_str ();
+  }
+
+  void mmon_subcomponent::add_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_add (size);
+  }
+
+  void mmon_subcomponent::sub_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_sub (size);
+  }
+
+  const char *mmon_component::get_name ()
+  {
+    return m_comp_name.c_str ();
+  }
+
+  void mmon_component::add_stat (uint64_t size, int subcomp_idx, bool init, bool expand)
+  {
+    /* description of add_stat(..).
+     * 1) if init == true, add init_stat
+     * 2) then, add cur_stat
+     * 3) compare with peak_stat
+     * 4) if cur_stat(new) > peak_stat, update peak_stat
+     * 5) if expand == true, expand_count++
+     * 6) call subcomponent->add_cur_stat(size) */
+    return;
+  }
+
+  void mmon_component::sub_stat (uint64_t size, int subcomp_idx, bool init)
+  {
+    /* description of sub_stat(size, init).
+     * 1) if init == true, sub init_stat
+     * 2) then, sub cur_stat
+     * 3) call component->sub_stat(size, subcomp_idx, init) */
+    return;
+  }
+
+  int mmon_component::is_subcomp_exist (const char *subcomp_name)
+  {
+    for (int i = 0; i < m_subcomponent.size (); i++)
+      {
+	if (!strcmp (subcomp_name, m_subcomponent[i]->get_name()))
+	  {
+	    return i;
+	  }
+      }
+
+    /* if not exist, return mmon_module::MAX_COMP_IDX */
+    return mmon_module::MAX_COMP_IDX;
+  }
+
+  int mmon_component::add_subcomponent (const char *name)
+  {
+    return 0;
+  }
+
+  mmon_module::mmon_module (const char *module_name, const MMON_COMP_INFO *info)
+    : m_module_name (module_name)
+  {
+    /* register component and subcomponent information
+     * add component and subcomponent */
+    int cnt = 0;
+    while (info[cnt].id != MMON_STAT_LAST)
+      {
+	bool comp_skip = false;
+	bool subcomp_skip = false;
+	int comp_idx = mmon_module::MAX_COMP_IDX, subcomp_idx = mmon_module::MAX_COMP_IDX;
+	int i;
+	if (info[cnt].comp_name)
+	  {
+	    for (i = 0; i < m_component.size (); i++)
+	      {
+		if (!strcmp (info[cnt].comp_name, m_component[i]->get_name ()))
+		  {
+		    comp_idx = i;
+		    comp_skip = true;
+		    break;
+		  }
+	      }
+
+	    if (!comp_skip)
+	      {
+		comp_idx = add_component (info[cnt].comp_name);
+	      }
+
+	    if (info[cnt].subcomp_name)
+	      {
+		subcomp_idx = m_component[comp_idx]->is_subcomp_exist (info[cnt].subcomp_name);
+		if (subcomp_idx != mmon_module::MAX_COMP_IDX)
+		  {
+		    subcomp_skip = true;
+		  }
+
+		if (!subcomp_skip)
+		  {
+		    subcomp_idx = m_component[comp_idx]->add_subcomponent (info[cnt].subcomp_name);
+		  }
+	      }
+	  }
+
+	m_comp_idx_map.push_back (MAKE_STAT_IDX_MAP (comp_idx, subcomp_idx));
+
+	cnt++;
+      }
+  }
+
+  int mmon_module::aggregate_stats (const MMON_MODULE_INFO &info)
+  {
+    return 0;
+  }
+
+  void mmon_module::add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init, bool expand)
```

Comment:
I did not see the parameters `bool init` and `bool expand` in the detailed design document. 
They are likely used for the situation when setting initial memory or reallocating memory. 

It seems that these parameters should be renamed to more meaningful names.

Reply:
There were many aspects related to C++ that were not considered during the detailed design (e.g., the use of private) and as a result, there were changes in the design. It seems that there needs to be a discussion on whether to modify the document accordingly or simply leave it as a description in the story.

Due to the movement of member variables like vectors to the private section, in the top-level class (memory_monitor class), access to deeper levels and member variables of member variables became inaccessible. As a result, the structure was changed so that each class would request updates to their own member variables for the increment and decrement of stats.

The "init" and "expand" mentioned in the comments are additional parameters that arose from this process. They are necessary parameters to determine whether to increment init_stat along with cur_stat and whether to increase expand_count.

Since they are hidden from actual users, I think it's not a big problem.

Reply:
Thank you for the detailed explanation. 

However, I still think we should change the parameter names. This is to ensure that any other maintainer besides you can easily understand and modify this code in the future. 

Additionally, is there a possibility that the below description comment will be maintained in the future or that function comment (args, return value, ...) will be added?

If not, I believe it would be better to rename the functions within a reasonably understandable scope. For example:

init -> increase_init_stat
expand -> increase_expand_count

Note:
The parameter names used in functions that receive boolean values are as follows:

`void recovery_analysis(..., bool skip_empty_transactions);`
`void load_checkpoint_trans(..., bool &at_least_one_active_transaction_has_valid_mvccid);`
`void disconnect_async(bool with_disc_msg);`

Reply:
I stand negative side of using long variable and parameter names, but I agree that the names you suggested may not feel particularly long. Let's discuss this topic again in the future.

By the way, how do you think about changing
init -> allocation_at_init
expand -> is_resized?
While the boolean parameter makes the actions of increasing init_stat and expand_count, I believe it is inappropriate to use the name of a boolean parameter as the name for the actual actions to be performed.

If you have any different opinions or better suggestions, please comment and let me know.